glide-core/redis-rs/redis/src/cluster_async/mod.rs

```diff
@@ -2139,6 +2204,319 @@ where
             .map_err(|err| (OperationTarget::Node { address }, err))
     }
 
+    // This function handles commands with routing info of MultiSlot (like MSET or MGET), creates sub-commands for the matching slots and add it to the correct pipeline
+    async fn handle_multi_slot_routing(
+        pipelines_by_connection: &mut NodePipelineMap<C>,
+        core: Core<C>,
+        cmd: &Cmd,
+        index: usize,
+        slots: Vec<(Route, Vec<usize>)>,
+    ) {
+        // inner_index is used to keep track of the index of the sub-command inside cmd
+        for (inner_index, (route, indices)) in slots.iter().enumerate() {
+            let conn = {
+                let lock = core.conn_lock.read().expect(MUTEX_READ_ERR);
+                lock.connection_for_route(route)
+            };
+            if let Some((address, conn)) = conn {
+                let new_cmd =
+                    crate::cluster_routing::command_for_multi_slot_indices(cmd, indices.iter());
+                add_command_to_node_pipeline(
+                    pipelines_by_connection,
+                    address,
+                    conn.await,
+                    new_cmd,
+                    index,
+                    Some(inner_index),
+                );
+            }
+        }
+    }
+
+    fn determine_internal_routing(
+        routing_info: SingleNodeRoutingInfo,
+    ) -> InternalSingleNodeRouting<C> {
+        match routing_info {
+            SingleNodeRoutingInfo::Random => SingleNodeRoutingInfo::Random.into(),
+            SingleNodeRoutingInfo::SpecificNode(route) => {
+                let route_single: SingleNodeRoutingInfo = Some(route).into();
+                route_single.into()
+            }
+            SingleNodeRoutingInfo::RandomPrimary => {
+                let route_single: SingleNodeRoutingInfo = Some(Route::new_random_primary()).into();
+                route_single.into()
+            }
+            SingleNodeRoutingInfo::ByAddress { host, port } => {
+                let address = format!("{host}:{port}");
+                InternalSingleNodeRouting::ByAddress(address)
+            }
+        }
+    }
+
+    // This function handles commands with routing info of SingleNode
+    async fn handle_single_node_route(
```

Comment:
same issue - all of these functions are placed under ClusterConnInner though they are only relevant for pipelines. I'm not sure there is a good reason placing them here. Why not moving all of the pipelines logic into a separate file (e.g. pipeline, pipeline_routing) under the async_cluster folder?